### PR TITLE
added unit tests for ProcessingPostIngestPlugin and InMemoryProcessingFramework

### PR DESCRIPTION
**Pull request by emmberk**
#### What does this PR do?
added unit tests for ProcessingPostIngestPlugin, InMemoryProcessingFramework, and ProcessResourceImpl
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@troymohl @peterhuffer 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@emmberk
#### How should this be tested? (List steps with links to updated documentation)
Confirm unit tests succeed
#### What are the relevant tickets?
https://github.com/emmberk/alliance/pull/5
#### Checklist:
- [ ] Documentation Updated
- [ ] Change Log Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/InMemoryProcessingFrameworkTest.groovy [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+    def 'test construct will null arguments'(CatalogFramework catalogFramework, ExecutorService threadPool) {
+        when:
+        new InMemoryProcessingFramework(catalogFramework, threadPool)
+
+        then:
+        thrown(NullPointerException)
+
+        where:
+        catalogFramework       | threadPool
```
Comment (troymohl) [inline catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/InMemoryProcessingFrameworkTest.groovy]: Aren't `catalogFramework` and `threadPool` both mocked and should be non-null?

File: catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/ProcessingPostIngestPluginTest.groovy [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+    def 'test construct with null arguments'(CatalogFramework catalogFramework, ProcessingFramework processingFramework) {
+        when:
+        new ProcessingPostIngestPlugin(catalogFramework, processingFramework)
+
+        then:
+        thrown(NullPointerException)
+
+        where:
+        catalogFramework       | processingFramework
```
Comment (troymohl) [inline catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/ProcessingPostIngestPluginTest.groovy]: Shouldn't both of these be non-null and not throw an NPE?

File: catalog/plugin/catalog-plugin-async/pom.xml [hunk 3]
```diff
@@ -100,22 +106,22 @@
...
                                             <counter>BRANCH</counter>
                                             <value>COVEREDRATIO</value>
-                                            <minimum>0.0</minimum>
+                                            <minimum>0.75</minimum>
                                         </limit>
                                         <limit>
                                             <counter>COMPLEXITY</counter>
                                             <value>COVEREDRATIO</value>
-                                            <minimum>0.0</minimum>
+                                            <minimum>0.75</minimum>
```
Comment (emmberk) [inline catalog/plugin/catalog-plugin-async/pom.xml]: Reminder that we need to get this ratio up:
```
[WARNING] Rule violated for bundle DDF :: Catalog :: Core :: Async Plug-in: instructions covered ratio is 0.87, but expected minimum is 0.99
[WARNING] Rule violated for bundle DDF :: Catalog :: Core :: Async Plug-in: branches covered ratio is 0.89, but expected minimum is 0.99
[WARNING] Rule violated for bundle DDF :: Catalog :: Core :: Async Plug-in: complexity covered ratio is 0.73, but expected minimum is 0.99
[WARNING] Rule violated for bundle DDF :: Catalog :: Core :: Async Plug-in: lines covered ratio is 0.87, but expected minimum is 0.99
```

File: catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/api/impl/ProcessResourceImplTest.groovy [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            processResource.mimeTypeRawData == MIME_TYPE_RAW_DATA
+            processResource.getInputStream() == stream
+            processResource.id == ID
+            processResource.size == SIZE
+        }
+
+        where:
+        id | stream            | mimeType           | fileName  | size | qualifier | modified
+        ID | Mock(InputStream) | MIME_TYPE_RAW_DATA | FILE_NAME | SIZE | QUALIFIER | false
+        ID | Mock(InputStream) | null               | null      | SIZE | QUALIFIER | false
```
Comment (troymohl) [inline catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/api/impl/ProcessResourceImplTest.groovy]: Should we test QUALIFIER == null in this table?
Also should we test a null InputStream in one of the tests?
Might be worth testing null ID as well just to be complete.

File: catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/api/impl/ProcessResourceImplTest.groovy [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    def 'test process resource URISyntaxException'() {
```
Comment (troymohl) [inline catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/api/impl/ProcessResourceImplTest.groovy]: This test description seems inconsistent with what is being tested (IllegalArgumentException). Maybe create `where` section for this to test null as well as empty id?

File: catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/api/impl/ProcessResourceImplTest.groovy [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    def inputStream = Mock(InputStream)
+
+    def 'process resource no qualifier'() {
+        when:
+        ProcessResource processResource = new ProcessResourceImpl(ID, inputStream, MIME_TYPE_RAW_DATA, FILE_NAME, SIZE)
+
+        then:
+        processResource.getQualifier() == ''
+        processResource.isModified()
+        processResource.uri == ProcessResource.CONTENT_SCHEME + ":" + ID
```
Comment (troymohl) [inline catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/api/impl/ProcessResourceImplTest.groovy]: Maybe worth creating a constant for `ProcessResource.CONTENT_SCHEME + ":" + ID` since it's referenced multiple times

File: catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/api/impl/ProcessResourceImplTest.groovy [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            processResource.filename == FILE_NAME
+            processResource.mimeTypeRawData == MIME_TYPE_RAW_DATA
+            processResource.getInputStream() == stream
+            processResource.id == ID
+            processResource.size == SIZE
+        }
+
+        where:
+        id | stream            | mimeType           | fileName  | size | qualifier | modified
+        ID | Mock(InputStream) | MIME_TYPE_RAW_DATA | FILE_NAME | SIZE | QUALIFIER | false
```
Comment (peterhuffer) [inline catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/api/impl/ProcessResourceImplTest.groovy]: TODO: Fix this. Mock not working correctly in `where`

File: catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/api/impl/ProcessResourceImplTest.groovy [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        then:
+        thrown IllegalArgumentException
+    }
+
+    def 'test mark unmodified resource as modified'() {
+        when:
+        ProcessResource processResource = new ProcessResourceImpl(ID, inputStream, MIME_TYPE_RAW_DATA, FILE_NAME, SIZE, false)
+
+        then:
+        !processResource.modified
```
Comment (emmberk) [inline catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/api/impl/ProcessResourceImplTest.groovy]: I think that we should always test `processResource.isModified()` in this class instead of accessing `processResource.modified` directly

File: catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/api/impl/ProcessResourceImplTest.groovy [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        processResource.markAsModified()
```
Comment (emmberk) [inline catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/api/impl/ProcessResourceImplTest.groovy]: This would read better with another `when:`/`then:` block:
```groovy
        when:
        def processResource = new ProcessResourceImpl(ID, inputStream, MIME_TYPE_RAW_DATA, FILE_NAME, SIZE, false)

        then:
        !processResource.isModified()

        when:
        processResource.markAsModified()

        then:
        processResource.isModified()
```

File: catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/api/impl/ProcessResourceImplTest.groovy (hunk 4 above)
Comment (peterhuffer) [inline catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/api/impl/ProcessResourceImplTest.groovy]: I didn't test null of InputStream or ID because they are assigned with no logic. I will add a qualifier with a null value

File: catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/ProcessingPostIngestPluginTest.groovy [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+import ddf.catalog.resource.ResourceNotFoundException
+import ddf.catalog.resource.ResourceNotSupportedException
+import org.codice.ddf.catalog.plugin.async.api.internal.ProcessCreateRequest
+import org.codice.ddf.catalog.plugin.async.api.internal.ProcessDeleteRequest
+import org.codice.ddf.catalog.plugin.async.api.internal.ProcessUpdateRequest
+import org.codice.ddf.catalog.plugin.async.api.internal.ProcessingFramework
+import spock.lang.Specification
+
+class ProcessingPostIngestPluginTest extends Specification {
+
```
Comment (troymohl) [inline catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/ProcessingPostIngestPluginTest.groovy]: Should we have a test that validates that a plugin makes a change to the metacard?

File: catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/ProcessingPostIngestPluginTest.groovy [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+        then:
+        output == input
+
+        1 * catalogFramework.getResource(_ as ResourceRequest, sourceId) >> {
+            throw exception
+        }
+
+        1 * processingFramework.submit(_ as ProcessCreateRequest) >> { ProcessCreateRequest processCreateRequest ->
+            processCreateRequest.getProcessCreateItems().get(0).getProcessResource() == null
```
Comment (peterhuffer) [inline catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/ProcessingPostIngestPluginTest.groovy]: Do these lines need the `assert` here?

Comment (emmberk) [inline catalog/plugin/catalog-plugin-async/src/test/groovy/org/codice/ddf/catalog/plugin/async/ProcessingPostIngestPluginTest.groovy]: good catch 👍
